src/adapters/methods/reft.py

```diff
@@ -0,0 +1,184 @@
+                        hidden_states[i, self.prefix_positions : suff_id, :],
+                        adapted_output[i, self.prefix_positions :, :],
+                        hidden_states[i, suff_id + self.suffix_positions :, :],
+                    ],
+                    dim=0,
+                )
+            )
+
+        output = torch.stack(output, dim=0)
+
+        return output
+
+    def forward(self, hidden_states: torch.Tensor):
+        # hidden_states shape: (batch_size, seq_len, hidden_size)
+        adapted_states, suff_ids = self._gather_adapted_states(hidden_states)
+
+        # apply reft
+        for i, unit in enumerate(self.units):
+            adapted_states[i] = unit(adapted_states[i])
+
+        output = self._scatter_adapted_states(hidden_states, adapted_states, suff_ids)
+
+        return output
+
+
+class ReftLayer(AdapterLayerBase, nn.Module):
+    adapter_modules_name = "refts"
+
+    def __init__(self, model_config, adapters_config):
+        super().__init__()
+        self.location_key = "reft"
+        self.model_config = model_config
+        self.adapters_config = adapters_config
+        self.refts = nn.ModuleDict()
+
+    def add_adapter(self, adapter_name: str, layer_idx: int) -> bool:
+        self.layer_idx = layer_idx
+        reft_config = self.adapters_config.match(
+            adapter_name,
+            config_type=ReftConfig,
+            layer_idx=self.layer_idx,
+        )
+        if reft_config is not None and (reft_config.layers == "all" or self.layer_idx in reft_config.layers):
+            reft = ReftModule(
+                self.model_config.hidden_size,
+                reft_config,
+            )
+            reft.train(self.training)
+            self.refts[adapter_name] = reft
+            return True
+
+        return False
+
+    def forward(self, hidden_states: torch.Tensor):
+        adapter_setup = self.get_active_setup()
+        if adapter_setup is not None:
+            first_adapter = adapter_setup.first()
+            if first_adapter in self.refts:
+                hidden_states = self.refts[first_adapter](hidden_states)
+
+        return hidden_states
+
+    def pre_save_adapters(self):
+        # Make sure orthogonal parametrizations are contiguous, otherwise saving with safetensors will fail
+        for reft in self.refts.values():
+            for unit in reft.units:
+                if unit.orthogonal:
+                    unit.projection.parametrizations.weight[0].base = unit.projection.parametrizations.weight[
+                        0
+                    ].base.contiguous()
+
+
+def hook_fn(module, args, output):
+    if isinstance(output, torch.Tensor):
+        return module.reft_layer(output)
+    else:
+        return (module.reft_layer(output[0]),) + output[1:]
+
+
+def init_reft(model):
+    for _, layer in model.iter_layers():
+        if not hasattr(layer, "reft_layer"):
+            layer.reft_layer = ReftLayer(model.config, model.adapters_config)
```

Comment:
Where is the `reft_layer` attribute added to the modules?

Reply:
It's added in this line and nowhere else. Is there something to change about this logic?

Reply:
Ah nevermind, I have got the logic now